osquery/database/plugins/rocksdb.cpp

```diff
@@ -114,6 +120,41 @@ class RocksDBDatabasePlugin : public DatabasePlugin {
 /// Backing-storage provider for osquery internal/core.
 REGISTER_INTERNAL(RocksDBDatabasePlugin, "database", "rocksdb");
 
+class RocksDBGuard {
+ public:
+  RocksDBGuard() : my_tid_(platformGetTid()), locked_(false) {}
+
+  bool safeEnter() {
+    if (!owner_tid_is_valid_ || owner_tid_ != my_tid_) {
+      mu_.lock();
```

Comment:
So, would it be acceptable to do something along the following lines:

Alternative 1:
GLogRocksDBLogger::Logv(...) {
    Guard_begin()
    BufferedLogSink::forward(false);           // force BufferedLogSink into "buffering" mode.
    LOG(INFO) << log_line;                        // This will never cause a re-entrant call anymore
    BufferedLogSink::forward(true);            // return BufferedLogSink to normal mode
    Guard_exit()
}

Then, modify BufferedLogSink::send() to check the buffer if forward_ is true.

The side effect of this is that some LOG calls in unrelated threads might end up being buffered for an instant of time, while Logv has forward turned off.

The "elegance" here is that there's no lock/guard/complicated mechanism needed at all. We use existing classes to meet the non-reentrancy requirement of rocksdb.

The "hack" here is that we are using a mechanism that was clearly intended to be used during initialization. In particular, the forward_ member of BufferedLogSink is not protected, so, in theory, some other thread could manipulate forward_

Alternative 2:

This alternative is even simpler, but a bit of a hack. In BufferedLogSink::send(), we use the initial part of the message to deduce whether the message is from GlogRocksDBLogger::Logv. If so, we just buffer, and never forward. If the initial part of the message is *not* from Logv, then, in addition to executing the normal code, we drain anything that has been buffered up.